fog/view/server/src/block_tracker.rs

```diff
@@ -464,344 +515,790 @@ mod tests {
+            status: IngressPublicKeyStatus {
+                start_block: 10,
+                pubkey_expiry: 17,
+                retired: true,
+                lost: false,
+            },
+            last_scanned_block: None,
         };
 
-        let missing_ranges = vec![BlockRange::new(0, ingestable_range.start_block)];
+        for i in 0..7 {
+            let index = rec1.status.start_block + i;
+
+            // Check that next blocks value matches what we expect
+            let expected_state = HashMap::from_iter(vec![(rec1.key, index)]);
+
+            assert_eq!(
+                block_tracker.next_blocks(&[rec1.clone()]),
+                expected_state,
+                "i = {}",
+                i
+            );
+
+            rec1.last_scanned_block = Some(index);
+            block_tracker.block_processed(rec1.key, index);
+
+            // Check that highest fully processed block count matches what we expect
+            let expected = (index + 1, None);
+            assert_eq!(
+                block_tracker.highest_fully_processed_block_count(&[rec1.clone()]),
+                expected,
+                "i = {}",
+                i
+            );
+        }
+
+        // Check that next blocks value matches what we expect
+        let expected_state = HashMap::from_iter(vec![]);
+
+        assert_eq!(block_tracker.next_blocks(&[rec1.clone()]), expected_state);
+
+        // Check that highest fully processed block count matches what we expect
+        let expected = (17, None);
+        assert_eq!(
+            block_tracker.highest_fully_processed_block_count(&[rec1.clone()]),
+            expected
+        );
+
+        // Now add a new key that comes much later and check that we can make progress
+        let rec2 = IngressPublicKeyRecord {
+            key: CompressedRistrettoPublic::from_random(&mut rng),
+            status: IngressPublicKeyStatus {
+                start_block: 30,
+                pubkey_expiry: 47,
+                retired: false,
+                lost: false,
+            },
+            last_scanned_block: None,
+        };
+
+        for i in 0..7 {
+            let index = rec2.status.start_block + i;
+
+            // Check that next blocks value matches what we expect
+            let expected_state = HashMap::from_iter(vec![(rec2.key, index)]);
 
-        for i in 0..10 {
             assert_eq!(
-                block_tracker.highest_fully_processed_block_count(
-                    &[ingestable_range.clone()],
-                    &missing_ranges
-                ),
-                ingestable_range.start_block + i,
+                block_tracker.next_blocks(&[rec2.clone()]),
+                expected_state,
+                "i = {}",
+                i
             );
 
-            block_tracker.block_processed(ingestable_range.id, ingestable_range.start_block + i);
+            // Make the block "exist" and "load" it
+            rec1.last_scanned_block = Some(index);
```

Comment:
HMMMM why is this changing `rec1`? Doesn't seem like `rec1` is used so this is probably some leftover